rp2040-hal/src/xosc.rs

```diff
@@ -0,0 +1,192 @@
+    convert::Infallible,
+    ops::RangeInclusive
+};
+
+use embedded_time::{
+    fraction::Fraction,
+    fixed_point::FixedPoint,
+    rate::{
+        Hertz,
+        Megahertz,
+    },
+    duration::{
+        Seconds,
+        Milliseconds,
+        Duration
+    }
+};
+
+use nb::Error::WouldBlock;
+
+/// State of the Crystal Oscillator (typestate trait)
+pub trait State {}
+
+/// XOSC is disabled (typestate)
+pub struct Disabled;
+
+/// XOSC is initialized, ie we've given parameters (typestate)
+pub struct Initialized {
+    freq_hz: Hertz<u64>
+}
+
+/// Stable state (typestate)
+pub struct Stable{
+    freq_hz: Hertz<u64>
+}
+
+/// XOSC is disabling (typestate)
+pub struct Disabling;
+
+/// XOSC is in dormant mode (see Chapter 2, Section 16, §5)
+pub struct Dormant;
+
+impl State for Disabled {}
+impl State for Initialized {}
+impl State for Stable {}
+impl State for Disabling {}
+impl State for Dormant {}
+
+/// Possible errors when initializing the CrystalOscillator
+pub enum Error {
+    /// Frequency is out of the 1-15MHz range (see datasheet)
+    FrequencyOutOfRange,
+
+    /// Argument is bad : overflows, ...
+    BadArgument
+}
+
+/// A Crystal Oscillator.
+pub struct CrystalOscillator<S: State> {
+    device: rp2040_pac::XOSC,
+    state: S
+}
+
+impl<S: State> CrystalOscillator<S> {
+    /// Transitions the oscillator to another state.
+    fn transition<To: State>(self, state: To) -> CrystalOscillator<To> {
+        CrystalOscillator {
+            device: self.device,
+            state: state
+        }
+    }
+
+    /// Releases the underlying device.
+    pub fn free(self) -> rp2040_pac::XOSC {
+        self.device
+    }
+}
+
+impl CrystalOscillator<Disabled> {
+
+    /// Creates a new CrystalOscillator from the underlying device.
+    pub fn new(dev: rp2040_pac::XOSC) -> Self {
+        CrystalOscillator {
+            device: dev,
+            state: Disabled
+        }
+    }
+
+    /// Initializes the XOSC : frequency range is set, startup delay is calculated and set.
+    pub fn initialize(self, frequency: Megahertz<u32>) -> Result<CrystalOscillator<Initialized>, Error> {
+
+        const ALLOWED_FREQUENCY_RANGE: RangeInclusive<Megahertz<u32>> = Megahertz(1)..=Megahertz(15);
+        const STABLE_DELAY: Milliseconds<u64> = Milliseconds(1_u64);
+        const DIVIDER: Fraction = Fraction::new(1, 256);
+
+        if !ALLOWED_FREQUENCY_RANGE.contains(&frequency) {
+            return Err(Error::FrequencyOutOfRange)
+        }
+
+        self.device.ctrl.write(|w| {
+            w.freq_range()._1_15mhz();
+            w
+        });
+
+        let freq_hz: Hertz<u64> = frequency.into();
+        let delay_sec: Seconds<u64> = STABLE_DELAY.into();
+
+        //let startup_delay = ((freq_hz / 1000) + 128) / 256;
+        let startup_delay = delay_sec.
+            checked_mul(freq_hz.integer()).and_then(|r|
+                 r.to_generic::<u64>(DIVIDER).ok()
+             ).
+            ok_or(Error::BadArgument)?;
```

Comment:
Might be overkill but maybe we should add a check if `startup_delay.integer()` fits into a `u16`.

Maybe something like
```
            ok_or(Error::BadArgument)?
            .integer()
            .try_into()
            .ok_ok(Error::BadArgument)?;
```

Reply:
I kept pondering if using embedded_time types was actually a good idea because the conversion traits don't work for all core integer types and that brings "ugly" casts everywhere. Your suggestion might actually be a good way to make them look better.

Reply:
Related to that code is the use of u64. I was wondering, for the sake of low-level/embedded, should we constraint all integer types to u32/usize ?

Reply:
> Related to that code is the use of u64. I was wondering, for the sake of low-level/embedded, should we constraint all integer types to u32/usize ?

I think for code that mostly runs once during an initialization it shouldn't be a big deal. In this case u32 might suffice though it would allow us to represetz up to 4GHz in Hz.

Reply:
> I kept pondering if using embedded_time types was actually a good idea [...]

Yes I had the same struggle, it seems like a nice way to add all the different units, but I ended up in generics-hell because I wanted to take *any* rate. But maybe just sticking to the smallest common unit makes it a bit easier to handle.